File: scripts/migrate_gym_codes.dart
// scripts/migrate_gym_codes.dart
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:tapem/firebase_options_dev.dart';
import 'package:tapem/features/gym/domain/services/gym_code_service.dart';

/// Migration script to convert static gym codes to rotating codes
/// 
/// This script:
/// 1. Reads all gyms with existing 'code' field
/// 2. Creates initial rotating codes in gym_codes collection
/// 3. Sets expiration to 1 month from now
/// 4. Optionally removes old 'code' field from gyms
/// 
/// Usage: dart run scripts/migrate_gym_codes.dart [--remove-old-field]

Future<void> main(List<String> args) async {
  final removeOldField = args.contains('--remove-old-field');
  
  print('🚀 Starting Gym Code Migration...\n');
  
  // Initialize Firebase
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );
  
  final firestore = FirebaseFirestore.instance;
  final gymCodeService = GymCodeService(firestore: firestore);
  
  try {
    // Get all gyms
    print('📋 Fetching all gyms...');
    final gymsSnapshot = await firestore.collection('gyms').get();
    print('Found ${gymsSnapshot.docs.length} gyms\n');
    
    var successCount = 0;
    var skipCount = 0;
    var errorCount = 0;
    
    for (final gymDoc in gymsSnapshot.docs) {
      final gymId = gymDoc.id;
      final gymData = gymDoc.data();
      final gymName = gymData['name'] as String? ?? 'Unknown';
      final oldCode = gymData['code'] as String?;
      
      print('Processing: $gymName ($gymId)');
      
      // Check if gym already has rotating codes
      final existingCode = await gymCodeService.getActiveCodeForGym(gymId);
      if (existingCode != null) {
        print('  ⏭️  Already has rotating code: ${existingCode.code}');
        print('  Expires: ${existingCode.expiresAt}');
        skipCount++;
        print('');
        continue;
      }
      
      try {
        // Create code document with unique code
        final gymCode = await gymCodeService.createCode(
          gymId: gymId,
          createdBy: 'migration-script',
          expiresAt: DateTime.now().add(const Duration(days: 30)),
        );
        
        print('  ✅ Created rotating code: ${gymCode.code}');
        print('  Expires: ${gymCode.expiresAt}');
        
        if (oldCode != null) {
          print('  Old static code was: $oldCode');
          
          if (removeOldField) {
            await gymDoc.reference.update({
              'code': FieldValue.delete(),
            });
            print('  🗑️  Removed old code field');
          }
        }
        
        successCount++;
      } catch (e) {
        print('  ❌ Error: $e');
        errorCount++;
      }
      
      print('');
    }
    
    // Summary
    print('━' * 50);
    print('Migration Complete!\n');
    print('✅ Successfully migrated: $successCount gyms');
    print('⏭️  Skipped (already migrated): $skipCount gyms');
    if (errorCount > 0) {
      print('❌ Errors: $errorCount gyms');
    }
    print('━' * 50);
    
    if (!removeOldField && successCount > 0) {
      print('\n💡 Tip: Run with --remove-old-field to remove old code fields');
    }
    
  } catch (e, stackTrace) {
    print('❌ Migration failed: $e');
    print(stackTrace);
    exit(1);
  }
  
  exit(0);
}

File: scripts/migrate_gym_codes.js
#!/usr/bin/env node

// scripts/migrate_gym_codes.js
// Simple Node.js script to migrate gym codes without Flutter dependencies

const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');

// Load service account from firebase.json to determine which project
const firebaseJson = JSON.parse(fs.readFileSync(path.join(__dirname, '../firebase.json'), 'utf8'));
const projectId = firebaseJson.projects?.default || 'tap-em-dev';

console.log(`🚀 Starting Gym Code Migration for ${projectId}...\n`);

// Initialize Firebase Admin
admin.initializeApp({
    projectId: projectId
});

const db = admin.firestore();

// Generate readable 6-character code
function generateCode() {
    const chars = 'ABCDEFGHJKLMNPQRTUVWXY3468';
    let code = '';
    for (let i = 0; i < 6; i++) {
        code += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return code;
}

// Check if code exists
async function codeExists(code) {
    const snapshot = await db.collectionGroup('codes')
        .where('code', '==', code)
        .limit(1)
        .get();
    return !snapshot.empty;
}

// Generate unique code
async function generateUniqueCode(maxAttempts = 10) {
    for (let i = 0; i < maxAttempts; i++) {
        const code = generateCode();
        const exists = await codeExists(code);
        if (!exists) {
            return code;
        }
    }
    throw new Error(`Failed to generate unique code after ${maxAttempts} attempts`);
}

async function migrate() {
    try {
        // Get all gyms
        console.log('📋 Fetching all gyms...');
        const gymsSnapshot = await db.collection('gyms').get();
        console.log(`Found ${gymsSnapshot.size} gyms\n`);

        let successCount = 0;
        let skipCount = 0;
        let errorCount = 0;

        for (const gymDoc of gymsSnapshot.docs) {
            const gymId = gymDoc.id;
            const gymData = gymDoc.data();
            const gymName = gymData.name || 'Unknown';
            const oldCode = gymData.code;

            console.log(`Processing: ${gymName} (${gymId})`);

            // Check if gym already has rotating codes
            const existingCodeSnapshot = await db
                .collection('gym_codes')
                .doc(gymId)
                .collection('codes')
                .where('isActive', '==', true)
                .where('expiresAt', '>', admin.firestore.Timestamp.now())
                .limit(1)
                .get();

            if (!existingCodeSnapshot.empty) {
                const existingCode = existingCodeSnapshot.docs[0].data();
                console.log(`  ⏭️  Already has rotating code: ${existingCode.code}`);
                console.log(`  Expires: ${existingCode.expiresAt.toDate().toISOString()}`);
                skipCount++;
                console.log('');
                continue;
            }

            try {
                // Generate unique code
                const newCode = await generateUniqueCode();

                // Calculate expiration (30 days from now)
                const now = new Date();
                const expiresAt = new Date(now);
                expiresAt.setDate(expiresAt.getDate() + 30);

                // Create code document
                await db
                    .collection('gym_codes')
                    .doc(gymId)
                    .collection('codes')
                    .add({
                        code: newCode,
                        gymId: gymId,
                        createdAt: admin.firestore.Timestamp.now(),
                        expiresAt: admin.firestore.Timestamp.fromDate(expiresAt),
                        isActive: true,
                        createdBy: 'migration-script'
                    });

                console.log(`  ✅ Created rotating code: ${newCode}`);
                console.log(`  Expires: ${expiresAt.toISOString()}`);

                if (oldCode) {
                    console.log(`  Old static code was: ${oldCode}`);
                }

                successCount++;
            } catch (error) {
                console.error(`  ❌ Error: ${error.message}`);
                errorCount++;
            }

            console.log('');
        }

        // Summary
        console.log('━'.repeat(50));
        console.log('Migration Complete!\n');
        console.log(`✅ Successfully migrated: ${successCount} gyms`);
        console.log(`⏭️  Skipped (already migrated): ${skipCount} gyms`);
        if (errorCount > 0) {
            console.log(`❌ Errors: ${errorCount} gyms`);
        }
        console.log('━'.repeat(50));

        process.exit(0);
    } catch (error) {
        console.error('❌ Migration failed:', error);
        process.exit(1);
    }
}

migrate();
